sympy/core/operations.py

```diff
@@ -44,13 +44,13 @@ class AssocOp(Basic):
     _args_type = None
 
     @cacheit
-    def __new__(cls, *args, **options):
+    def __new__(cls, *args, evaluate=None, _sympify=True):
         from sympy import Order
 
         # Allow faster processing by passing ``_sympify=False``, if all arguments
         # are already sympified.
-        if options.get('_sympify', True):
-            args = list(map(_sympify, args))
+        if _sympify:
+            args = list(map(_sympify_, args))
```

Comment:
I wonder if this would make more sense as:
```
    def __new__(cls, *args, evaluate=None):
        return cls._new_from_basic(*map(_sympify_, args))

    @cacheit  # note: only cache post-sympification
    def _new_from_basic(cls, *args, evaluate=None):
        from sympy import Order
```

That avoids the need for a "private" argument, which is usually an antipattern

Reply:
The purpose of `_sympify` argument is to skip sympifying step if we are certain that `args` are already sympified. Did you intentionally remove `cachit` from `__new__` in your suggestion? It seems that `map(_sympify_, args)` will be run every time.

Reply:
I think it's better to keep the cache on `__new__`. The performance sensitive paths in sympy are those where everything is already sympified and avoiding resympification is worthwhile particularly in the case of a cache hit which is otherwise a very fast operation.

Reply:
> The purpose of _sympify argument is to skip sympifying step if we are certain that args are already sympified.

Right, I'm advocating `Add._new_from_basic(already, sympified)` instead of `Add(already, sympified, _sympify=False)` because it avoids needing the private argument

> The performance sensitive paths in sympy are those where everything is already sympified

In principle those paths could be changed to call `_new_from_basic`, but perhaps that's not worth the effort.

> particularly in the case of a cache hit which is otherwise a very fast operation.

Is `sympify` not also a very fast operation for already-sympified objects?

Reply:
Sympify for sympified objects is already reduced to a single attribute lookup (apart from the overhead of two function calls, and the attribute is a property). The thing is that here we have to call sympify for each object. Where as cacheit can cache the whole tuple.
```julia
In [15]: @cacheit 
    ...: def cnew(*objs): 
    ...:     return sum(map(sympify, objs)) 
    ...:                                                                                                                                       

In [16]: def new(*objs): 
    ...:     return csum(map(sympify, objs)) 
    ...:                                                                                                                                       

In [17]: @cacheit 
    ...: def csum(objs): 
    ...:     return sum(objs) 
    ...:                                                                                                                                       

In [18]: %time cnew(*range(100))                                                                                                               
CPU times: user 737 µs, sys: 18 µs, total: 755 µs
Wall time: 866 µs
Out[18]: 4950

In [19]: %time cnew(*range(100))                                                                                                               
CPU times: user 19 µs, sys: 0 ns, total: 19 µs
Wall time: 24.1 µs
Out[19]: 4950

In [20]: %time new(*range(100))                                                                                                                
CPU times: user 474 µs, sys: 1 µs, total: 475 µs
Wall time: 479 µs
Out[20]: 4950

In [21]: %time new(*range(100))                                                                                                                
CPU times: user 474 µs, sys: 10 µs, total: 484 µs
Wall time: 489 µs

In [22]: d = list(map(Integer, range(100)))                                                                                                   

In [24]: %time cnew(*d)                                                                                                                        
CPU times: user 403 µs, sys: 33 µs, total: 436 µs
Wall time: 631 µs
Out[24]: 4950

In [25]: %time cnew(*d)                                                                                                                        
CPU times: user 54 µs, sys: 7 µs, total: 61 µs
Wall time: 65.8 µs
Out[25]: 4950

In [26]: %time new(*d)                                                                                                                         
CPU times: user 335 µs, sys: 3 µs, total: 338 µs
Wall time: 342 µs
Out[26]: 4950

In [27]: %time new(*d)                                                                                                                         
CPU times: user 336 µs, sys: 0 ns, total: 336 µs
Wall time: 342 µs
Out[27]: 4950
```

Reply:
I think that example is bad because `csum` is a cache miss every time, as `map` objects are not cacheable.

Reply:
Actually that test isn't a fair comparison because it should use `*args` for csum.

Reply:
I assume you are referring to your test and not mine?

Reply:
Yes, I was referring to mine (making the same point that you did, but our messages crossed paths). With that change I get:
```julia
In [61]: @cacheit 
    ...: def cnew(*objs): 
    ...:     return sum(map(sympify, objs)) 
    ...:  
    ...: def new(*objs): 
    ...:     return csum(*map(sympify, objs)) 
    ...:  
    ...: @cacheit 
    ...: def csum(*objs): 
    ...:     return sum(objs) 
    ...:                                                                                                                                       

In [62]: %time new(*range(100))                                                                                                                
CPU times: user 868 µs, sys: 805 µs, total: 1.67 ms
Wall time: 1.71 ms
Out[62]: 4950

In [63]: %time new(*range(100))                                                                                                                
CPU times: user 297 µs, sys: 1 µs, total: 298 µs
Wall time: 304 µs
Out[63]: 4950

In [64]: %time cnew(*range(100))                                                                                                               
CPU times: user 488 µs, sys: 1 µs, total: 489 µs
Wall time: 495 µs
Out[64]: 4950

In [65]: %time cnew(*range(100))                                                                                                               
CPU times: user 19 µs, sys: 0 ns, total: 19 µs
Wall time: 24.1 µs
Out[65]: 4950
```
The important comparison though is with the already sympified args. Those are slower on a cache-hit than builtin types because `__hash__` is slower:
```julia
In [4]: d = list(map(Integer, range(100)))                                                                                                     

In [5]: %time new(*d)                                                                                                                          
CPU times: user 503 µs, sys: 1 µs, total: 504 µs
Wall time: 511 µs
Out[5]: 4950

In [6]: %time new(*d)                                                                                                                          
CPU times: user 147 µs, sys: 1 µs, total: 148 µs
Wall time: 153 µs
Out[6]: 4950

In [7]: %time cnew(*d)                                                                                                                         
CPU times: user 310 µs, sys: 1 µs, total: 311 µs
Wall time: 315 µs
Out[7]: 4950

In [8]: %time cnew(*d)                                                                                                                         
CPU times: user 48 µs, sys: 1 µs, total: 49 µs
Wall time: 52 µs
Out[8]: 4950
```